core/evm_ext.go

```diff
@@ -0,0 +1,107 @@
+// (c) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package core
+
+import (
+	"github.com/ava-labs/avalanchego/snow"
+	"github.com/ava-labs/subnet-evm/constants"
+	"github.com/ava-labs/subnet-evm/core/vm"
+	"github.com/ava-labs/subnet-evm/params"
+	"github.com/ava-labs/subnet-evm/precompile/contract"
+	"github.com/ava-labs/subnet-evm/precompile/contracts/deployerallowlist"
+	"github.com/ava-labs/subnet-evm/precompile/modules"
+	"github.com/ava-labs/subnet-evm/precompile/precompileconfig"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+var (
+	newEVM = vm.NewEVM
+)
+
+func init() {
+	vm.NewEVM = NewEVM
+}
+
+// IsProhibited returns true if [addr] is in the prohibited list of addresses which should
+// not be allowed as an EOA or newly created contract address.
+func IsProhibited(addr common.Address) bool {
+	if addr == constants.BlackholeAddr {
+		return true
+	}
+
+	return modules.ReservedAddress(addr)
+}
+
+type EVM struct {
+	*vm.EVM
+
+	chainConfig *params.ChainConfig
+}
+
+func NewEVM(
+	blockCtx vm.BlockContext, txCtx vm.TxContext, statedb vm.StateDB,
+	chainConfig vm.ChainConfig, config vm.Config,
+) *vm.EVM {
+	customChainConfig, ok := chainConfig.(*params.ChainConfig)
+	if !ok {
+		// If the chainConfig is not a params.ChainConfig, then we can't use the custom
+		// EVM implementation, so we fall back to the default implementation.
+		log.Warn("ChainConfig is not a *params.ChainConfig, falling back to default EVM")
+		return newEVM(blockCtx, txCtx, statedb, chainConfig, config)
+	}
+	evm := &EVM{
+		chainConfig: customChainConfig,
+	}
+	config.IsProhibited = IsProhibited
+	config.DeployerAllowed = evm.DeployerAllowed
+	config.CustomPrecompiles = evm.CustomPrecompiles
+
+	evm.EVM = newEVM(blockCtx, txCtx, statedb, chainConfig, config)
+	return evm.EVM
+}
+
+func (evm *EVM) GetBlockContext() contract.BlockContext {
+	return &evm.EVM.Context
+}
+
+func (evm *EVM) GetStateDB() contract.StateDB {
+	return evm.StateDB
+}
+
+func (evm *EVM) GetChainConfig() precompileconfig.ChainConfig {
```

Comment:
Why is this being returned as a `precompileconfig.ChainConfig`?

My immediate thought, before inspecting the interface definition, was that it seems to be a non sequitur. We're in the `core` package, accepting something akin to a `params.ChainConfig`, and then it's converted into something to do with precompiles.

Upon inspection I see that it only has `IsDurango()`; as a rule of thumb, return types should be as broadly useful as possible and accepted interfaces should require the minimum-viable signature to function.

Reply:
It is already being returned as `precompileconfig.ChainConfig` but from `core/vm`, which seems worse.
Also this is the goal of this refactoring, to separate avalanche specifics out of the params package.

Probably you looked at coreth, it also includes `GetFeeConfig` and `AllowedFeeRecipients`.
IMO avalanche specifics should be given access to all the avalanche specific parameters, just it must be transparent to `core/vm` package, and later to the `core` package as well, as you note it is not ideal.